src/translator/response_builder.cpp

```diff
@@ -1,27 +1,28 @@
 #include "response_builder.h"
 #include "response_options.h"
+#include "quality_estimator.h"
 
 namespace marian {
 namespace bergamot {
 
+
 void ResponseBuilder::buildQualityScores(Histories &histories,
                                          Response &response) {
-  std::vector<Quality> qualityScores;
+  QualityEstimator model(qualityEstimator_->begin());
+  AnnotatedText source = response.source;
   for (auto &history : histories) {
-    // TODO(jerin): Change hardcode of nBest = 1
-    NBestList onebest = history->nBest(1);
-
-    Result result = onebest[0]; // Expecting only one result;
+    Result result = history->top(); // Expecting only one result;
     Words words = std::get<0>(result);
     auto hyp = std::get<1>(result);
+
     // Quality scores: Sequence level is obtained as normalized path scores.
     // Word level using hypothesis traceback. These are most-likely
     // logprobs.
     auto normalizedPathScore = std::get<2>(result);
     auto wordQualities = hyp->tracebackWordScores();
-    wordQualities.pop_back();
     response.qualityScores.push_back(
         Quality{normalizedPathScore, wordQualities});
+    model.compute_quality_scores(response, words);
```

Comment:
This needs to be optional.  Many (most?) translation requests will not want quality scores.

Reply:
Just to avoid confusion, the current convention is any `f` which gives an optional object (Alignments, QualityScores) does not handle the optional bit internally inside `f`, but is handled in `ResponseBuilder` which chooses to call for not depending on `ResponseOptions` contains it or not.

So effectively, this is already optional in that this method as a whole isn't called if `ResponseOptions.qualityScores = false`.